gcampus/core/tasks.py

```diff
@@ -0,0 +1,214 @@
+
+from celery import shared_task
+from django.conf import settings
+from django.core.mail import mail_managers
+from django.db import transaction
+from django.db.models import Q, Count
+from django.utils import timezone, translation
+
+from gcampus.auth.models import Course, AccessKey, CourseToken
+from gcampus.core.models import Measurement, Water
+from gcampus.mail.messages.maintenance import (
+    MaintenanceAccessKeys,
+    MaintenanceCourseDeletion,
+)
+
+logger = logging.getLogger("gcampus.core.tasks")
+
+
+@shared_task
+def send_course_deletion_email(
+    email: str,
+    course_name: Optional[str],
+    course_school: Optional[str],
+    language: Optional[str] = None,
+):
+    if language is None:
+        language = settings.LANGUAGE_CODE
+    translation.activate(language)
+
+    email_template = MaintenanceCourseDeletion(course_name, course_school)
+    message = email_template.as_message([email])
+    message.send()
+
+
+@shared_task
+def send_access_key_deactivation_email(
+    email: str,
+    course_name: Optional[str],
+    course_school: Optional[str],
+    access_keys: List[str],
+    language: Optional[str] = None,
+):
+    if language is None:
+        language = settings.LANGUAGE_CODE
+    translation.activate(language)
+
+    email_template = MaintenanceAccessKeys(course_name, course_school, access_keys)
+    message = email_template.as_message([email])
+    message.send()
+
+
+@shared_task
+def maintenance():
+    now = timezone.now()
+
+    measurements = Measurement.all_objects.filter(
+        Q(parameters__isnull=True),
+        # AND
+        Q(hidden=False),
+        # AND
+        Q(updated_at__lt=(now - settings.MEASUREMENT_RETENTION_TIME)),
+        # AND
+        Q(comment__isnull=True) | Q(comment__exact=""),
+    )
+    measurement_count: int = measurements.all().count()
+    measurements.update(hidden=True)
+
+    unverified_courses: List[Course] = Course.objects.filter(
+        email_verified=False,
+        updated_at__lt=(now - settings.UNVERIFIED_COURSE_RETENTION_TIME),
+    ).all()
+    unverified_courses_count = 0
+    with transaction.atomic():
+        for course in unverified_courses:
+            AccessKey.objects.filter(course=course).delete()
+            CourseToken.objects.filter(course=course).delete()
+            course.delete()
+            unverified_courses_count += 1
+
+    unused_courses: List[Course] = (
+        Course.objects.filter(
+            email_verified=True,
+            course_token__last_login__lt=(now - settings.UNUSED_COURSE_RETENTION_TIME),
+        )
+        .annotate(measurement_count=Count("access_keys__measurements"))
+        .filter(measurement_count=0)
+        .all()
+    )
+    unused_courses_count = 0
+    with transaction.atomic():
+        for course in unused_courses:
+            AccessKey.objects.filter(course=course).delete()
+            CourseToken.objects.filter(course=course).delete()
+            send_course_deletion_email.apply_async(
```

Comment:
If sending the email fails (due to an error on our side), should we skip deleting the course?

Reply:
The email is send in an extra task, meaning that it will be done later. Thus, we can not catch errors at this point. But I think that's okay because we only delete courses which have been inactive for 180 + 30 days and have no measurements. The likelihood of a course being this old, the email failing AND the teacher noticing the missing course is pretty slim.